---
title: "bioBakeryR: Run bioBakery workflow on Terra using R"
author: "Sehyun Oh"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  BiocStyle::html_document:
    toc: true
    toc_float: false
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo=FALSE}
suppressPackageStartupMessages({
  library(bioBakeryR)
  library(AnVIL)
})
source('~/data2/bioBakeryR/R/getOutput.R')
dest_dir <- "~/data2/bioBakeryR/inst/extdata/outputs"
HSM7J4NY_dir <- "~/data2/bioBakeryR/inst/extdata/outputs/HSM7J4NY"
data_dir <- "~/data2/bioBakeryR/inst/extdata/outputs/data"
```

# Overview
bioBakery workflows is a collection of workflows and tasks for executing common 
microbial community analyses using standardized, validated tools and parameters.
bioBakery is built and maintained by [Huttenhower lab](http://huttenhower.sph.harvard.edu/).
In this vignette, we demonstrate how to run [bioBakery workflow](https://github.com/biobakery/biobakery_workflows) 
on [Terra](https://app.terra.bio/#) using R package, bioBakeryR. 

**0. Prerequisite**   
You need [Terra account setup](https://support.terra.bio/hc/en-us/articles/360034677651-Account-setup-and-exploring-Terra). 
The email address linked to Terra account and the billing project name are required 
to use bioBakeryR package.   

**1. Setup**   
1-1. `cloneWorkspace` function copies the template workspace containing the bioBakery
workflow. Workspace is the main building block of [Terra](https://app.terra.bio/#), 
where different resources are delivered through. For bioBakeryR, the 'template' 
workspace is where default workflow is store and users will have their own copy 
of it, where they can modify, compute, and save their own results.   
1-2. `updateInput` function takes user's inputs. There are two required inputs: `ProjectName` 
(string) and `InputRead1Files` (tsv file). You can optionally provide metadata tsv file 
through `InputMetadataFile` argument. You can check the current input information using 
`currentInput` function.

**2. Run bioBakery workflow**   
`launchWorkflow` function launch the bioBakery workflow in Terra. You can abort 
the submission using `abortSubmission` function. For now, the input files should 
be saved in Google bucket. 

**3. Result**   
3-1. `monitorSubmission` function allows you to monitor the status of your workflow run.   
3-2. `listOutput` function displays the list of your workflow outputs.  
3-3. `tableHead` function allows you to check the head of output tsv files.   
3-4. `getOutput` function allows you to download your outputs.

<br>

Here are the basic input parameters used in this example.

```{r}
accountEmail = "shbrief@gmail.com"
billingProjectName = "waldronlab-terra-rstudio"
workspaceName = "mtx_workflow_biobakery_ver3"
```

# Setup
## Clone Workspace
```{r}
cloneWorkspace(accountEmail, billingProjectName, workspaceName)
```

You need to provide a unique name for the cloned workspace
```{r}
cloneWorkspace(accountEmail, billingProjectName, workspaceName = "test")
```

```{r echo=FALSE, message=FALSE}
## Delete test workspace
terra <- Terra()
resp <- terra$deleteWorkspace(workspaceNamespace = billingProjectName,
                              workspaceName = "test")
rm(resp)
```

## Format of inputs
`input` is a table containing the input file's Google bucket location. This
example file contains 6 fastq files.

```{r}
input <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_file_list.txt"
inputMeta <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_demo_metadata.txt"

read.table(gsutil_pipe(input), sep  = "\t")
```

`inputMeta` contains the metadata of the samples. 
```{r warning=FALSE}
read.table(gsutil_pipe(inputMeta), sep = "\t", header = TRUE)
```


## Update Input
Review the current input data.
```{r}
currentInput(accountEmail, billingProjectName, workspaceName)
```

Before launching the workflow, you should provide the correct input information 
using `updateInput` function.

```{r eval=FALSE}
## Still need to figure out the redundant argument, 'workspaceName'
input <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_file_list.txt"
inputMeta <- "gs://fc-07ee4ddc-5b5b-46f6-bed7-809aa14bb012/IBDMDB/ibdmdb_demo_metadata.txt"
updateInput(inputPath = input,
            inputMetadataPath = inputMeta,
            billingProjectName, workspaceName)
```





# Run bioBakery workflow
```{r}
launchWorkflow(accountEmail, billingProjectName, workspaceName)
```

## Monitor Progress
```{r}
recent_submission <- monitorSubmission(accountEmail, billingProjectName, workspaceName)
recent_submission
```

If you want to abort the submitted job, you can  
```{r}
recent_submissionId <- recent_submission$submissionId
abortSubmission(accountEmail, billingProjectName, workspaceName,
                submissionId = recent_submissionId)
```


# Result
## List Outputs (need to review)
For the demonstration purpose, we aborted the launched workflow above. This workspace
had a successful run with the same input already. To search that, we screened all
the previous submission status and found the one with 'Done' status.

```{r}
jobs <- monitorSubmission(accountEmail, billingProjectName, workspaceName,
                          mostRecentOnly = FALSE)
jobs_done <- which(jobs["status",]  == "Done")
jobs[,jobs_done]
```

To access the output of the submitted job, you need a submission Id.

```{r}
submission_id <- jobs[, jobs_done[1]]$submissionId
submission_id
```

List of the output files
```{r collapse=TRUE}
# This takes a while if you don't specify the keyword.
listOutput(accountEmail, billingProjectName, workspaceName, submissionId = submission_id)
listOutput(accountEmail, billingProjectName, workspaceName, submissionId = submission_id, 
           keyword = "humann")
```

Using `tableHead` function, you can check the head of output .tsv files without dowwnloading them.
```{r warning=FALSE, collapse=TRUE}
tableHead("HSM7J4NY_genefamilies_relab.tsv", n = 6,
          accountEmail, billingProjectName, workspaceName, 
          submissionId = submission_id)   # gene families relative abundance
tableHead("HSM7J4NY_pathcoverage.tsv", n = 6,
          accountEmail, billingProjectName, workspaceName, 
          submissionId = submission_id)   # pathcoverage
tableHead("HSM7J4NY.tsv", n = 6,
          accountEmail, billingProjectName, workspaceName, 
          submissionId = submission_id)   # metaphlan_bugs_list
```


## Get Outputs
### Sample-specific outputs
```{r}
listOutput(accountEmail, billingProjectName, workspaceName, submission_id, 
           keyword = "HSM7J4NY.*.tsv")
```

You can download any file using `getOutput` function.
```{r eval=FALSE}
HSM7J4NY_dir <- "~/data2/bioBakeryR/inst/extdata/outputs/HSM7J4NY"
getOutput(accountEmail, billingProjectName, workspaceName, submission_id, 
          keyword = "HSM7J4NY.*.tsv", dest_dir = HSM7J4NY_dir)
```

#### HSM7J4NY_genefamilies.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_genefamilies.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_pathabundance.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_pathabundance.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_pathcoverage.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_pathcoverage.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_ecs.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_ecs.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_kos.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_kos.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_ecs_relab.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_ecs_relab.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_genefamilies_relab.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_ecs_relab.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY_pathabundance_relab.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY_ecs_relab.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### HSM7J4NY.tsv
```{r echo=FALSE}
fpath <- file.path(HSM7J4NY_dir, "HSM7J4NY.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE,  skip =  3)
head(x, 3)
```




### Output files for visualization
```{r echo=FALSE, eval=FALSE}
listOutput(accountEmail, billingProjectName, workspaceName, submission_id, 
           keyword = "visualization")
```

Output files for visualization is saved in`ibdmdb_test_visualizations.zip`.
```{r eval=FALSE}
dest_dir <- "~/data2/bioBakeryR/inst/extdata/outputs"
getOutput(accountEmail, billingProjectName, workspaceName, submission_id, 
          keyword = "visualization", dest_dir = dest_dir)
unzip(file.path(dest_dir, "ibdmdb_test_visualizations.zip"), exdir = dest_dir)
```

```{r}
list.files(data_dir)
```

#### humann_feature_counts.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "humann_feature_counts.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### humann_read_and_species_count_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "humann_read_and_species_count_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### kneaddata_read_count_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "kneaddata_read_count_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### metaphlan_taxonomic_profiles.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "metaphlan_taxonomic_profiles.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### microbial_counts_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "microbial_counts_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### pathabundance_relab.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "pathabundance_relab.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### qc_counts_orphans_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "qc_counts_orphans_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### qc_counts_pairs_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "qc_counts_pairs_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### taxa_counts_table.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "taxa_counts_table.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```

#### top_average_pathways_names.tsv
```{r echo=FALSE}
fpath <- file.path(data_dir, "top_average_pathways_names.tsv")
x <- read.csv(fpath, sep = "\t", header = TRUE)
head(x, 3)
```




### Make SummarizedExperiment
#### Taxonomy profile
```{r}
taxo_profile <- read.csv(file.path(data_dir, "metaphlan_taxonomic_profiles.tsv"),
                         sep = "\t", header = TRUE)

## Taxonomy as rownames
taxo_profile <- tibble::column_to_rownames(taxo_profile, var = "X..taxonomy")

## Metadata
meta <- read.table("~/data2/bioBakeryR/inst/extdata/ibdmdb_demo_metadata.txt", 
                   sep = "\t", header = TRUE)
colData <- DataFrame(meta)
```

```{r}
se <- SummarizedExperiment(assays = list(taxo_profile = taxo_profile),
                           colData = colData)
se
```


#### Pathway abundance
```{r}
path_abun <- read.csv(file.path(data_dir, "pathabundance_relab.tsv"),
                      sep = "\t", header = TRUE)

## Pathway as rownames
path_abun <- tibble::column_to_rownames(path_abun, var = "X..Pathway")
colnames(path_abun) <- gsub("_Abundance", "", colnames(path_abun))

head(path_abun)
```

